### PR TITLE
Adds $purge-history operation

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
Adds support for `$purge-history` which can remove previous versions of a resource. This operation has been implemented as a variation of `hard-delete` since it is permanently removing data.

## Related issues
Addresses [issue #].

## Testing
Adds E2E tests

## FHIR Team Checklist
- [ ] **Update the title** of the PR to be succinct and less than 50 characters
- [ ] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [ ] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [ ] Tag the PR with **Azure API for FHIR** if this will release to the managed service
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch

### CODE REVIEW

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/DeleteTests.cs [hunk 1]
```diff
@@ -121,6 +121,34 @@ async Task ExecuteAndValidateNotFoundStatus(Func<Task> action)
...
+            string historyVersion = response.Resource.Meta.VersionId;
+
+            response.Resource.Meta.Tag.Add(new Coding("tests", Guid.NewGuid().ToString()));
+
+            using FhirResponse<Observation> updateResponse = await _client.UpdateAsync(response.Resource);
+
+            // Delete the resource.
+            await _client.DeleteAsync($"{nameof(Observation)}/{resourceId}/$purge");
+
+            // Subsequent read on the resource should return Gone.
```
Comment (LTA-Thinking) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/DeleteTests.cs:141]: This comment doesn't match the check that is happening after it. It looks like you are checking that the worked and the version id from a read matches the updated resource.

**Comment by LTA-Thinking** [comment]
There are tests for soft and hard delete in multiple test files. Do you think it is worthwhile to add purge to these test files as well?

File: src/Microsoft.Health.Fhir.Api/Features/Routing/KnownRoutes.cs [hunk 2]
```diff
@@ -58,6 +58,9 @@ internal class KnownRoutes
         public const string Everything = "$everything";
         public const string PatientEverythingById = "Patient/" + IdRouteSegment + "/" + Everything;
 
+        public const string Purge = "$purge";
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Api/Features/Routing/KnownRoutes.cs:61]: >$purge [](http://example.com/codeflow?start=37&length=6)

`PurgeHistory` maybe?
I would expect recent version to be deleted from purge as well.

Comment (CaitlinV39) [reply src/Microsoft.Health.Fhir.Api/Features/Routing/KnownRoutes.cs:61]: I really like this suggestion :) PurgeHistory makes more sense to me too.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs [hunk 3]
```diff
@@ -284,8 +284,13 @@ public async Task<ResourceWrapper> GetAsync(ResourceKey key, CancellationToken c
             }
         }
 
-        public async Task HardDeleteAsync(ResourceKey key, CancellationToken cancellationToken)
+        public async Task HardDeleteAsync(ResourceKey key, bool keepCurrentVersion, CancellationToken cancellationToken)
         {
+            if (keepCurrentVersion)
+            {
+                throw new NotImplementedException();
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs:291]: >NotImplementedException [](http://example.com/codeflow?start=26&length=23)

```
CREATE PROCEDURE dbo.HardDeleteResource
    @resourceTypeId smallint,
    @resourceId varchar(64),
    @keepCurrentVersion smallint,

AS
    SET NOCOUNT ON

    SET XACT_ABORT ON
    BEGIN TRANSACTION

    DECLARE @resourceSurrogateIds TABLE(ResourceSurrogateId bigint NOT NULL)

    DELETE FROM dbo.Resource
    OUTPUT deleted.ResourceSurrogateId
    INTO @resourceSurrogateIds
    WHERE ResourceTypeId = @resourceTypeId AND ResourceId = @resourceId 
    AND not(@keepCurrentVersion=1 and IsHistory=0)
```

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs [hunk 4]
```diff
@@ -8,11 +8,12 @@ namespace Microsoft.Health.Fhir.SqlServer.Features.Schema
...
     {
         public const int Min = (int)SchemaVersion.V4;
-        public const int Max = (int)SchemaVersion.V12;
+        public const int Max = (int)SchemaVersion.V13;
         public const int SearchParameterStatusSchemaVersion = (int)SchemaVersion.V6;
         public const int SupportForReferencesWithMissingTypeVersion = (int)SchemaVersion.V7;
         public const int SearchParameterHashSchemaVersion = (int)SchemaVersion.V8;
         public const int PartitionedTables = (int)SchemaVersion.V9;
         public const int SearchParameterSynchronizationVersion = (int)SchemaVersion.V12;
+        public const int PurgeHistoryVersion = (int)SchemaVersion.V12;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs:17]: >V12 [](http://example.com/codeflow?start=66&length=3)

V13?

**Comment by Ivanidzo4ka** [comment]
>     (12, 'started')

13?

---
Refers to: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/13.sql:9 in b8c8367. [](commit_id = b8c8367ea55e26f6e432f9e769f66e78af30bf47, deletion_comment = False)

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs (hunk 4 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/SchemaVersionConstants.cs]: You are right...

File: src/Microsoft.Health.Fhir.Api/Features/Routing/KnownRoutes.cs [hunk 5]
```diff
@@ -58,6 +58,9 @@ internal class KnownRoutes
         public const string Everything = "$everything";
         public const string PatientEverythingById = "Patient/" + IdRouteSegment + "/" + Everything;
 
+        public const string PurgeHistory = "$purge-history";
+        public const string PurgeHistoryResourceTypeById = ResourceTypeById + "/" + PurgeHistory;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Api/Features/Routing/KnownRoutes.cs:62]: >PurgeHistoryResourceTypeById [](http://example.com/codeflow?start=28&length=28)

To make it proper operations we need:
-Capability statement
-json file describing operation
-route which returns that json

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Api/Features/Routing/KnownRoutes.cs:62]: I think I mentioned that we should look at making operations a little simpler to add :)

**Review by Ivanidzo4ka — APPROVED** [review]
:shipit:

File: src/Microsoft.Health.Fhir.Api/Features/Routing/KnownRoutes.cs (hunk 5 above)
Comment (GitForceJedi) [reply src/Microsoft.Health.Fhir.Api/Features/Routing/KnownRoutes.cs:62]: Were they made simpler to add? What is the minimum required to add an operation, if all of the FHIR resources are available?

When asking the question I was pointed here. But this seems to have additional changes that may not be required.

**Pull request merged**